ml-agents/mlagents/trainers/torch/utils.py

```diff
@@ -15,6 +15,199 @@
+            )
+        if self.discrete_list is not None:
+            array_dict["discrete_action"] = ModelUtils.to_numpy(
+                self.discrete_tensor[:, 0, :]
+            )
+        return array_dict
+
+    def to_tensor_list(self) -> List[torch.Tensor]:
+        """
+        Returns the tensors in the AgentAction as a flat List of torch Tensors. This will be removed
+        when the ActionModel is merged.
+        """
+        tensor_list: List[torch.Tensor] = []
+        if self.continuous_tensor is not None:
+            tensor_list.append(self.continuous_tensor)
+        if self.discrete_list is not None:
+            tensor_list += (
+                self.discrete_list
+            )  # Note this is different for ActionLogProbs
+        return tensor_list
+
+    @staticmethod
+    def create(
+        tensor_list: List[torch.Tensor], action_spec: ActionSpec
+    ) -> "AgentAction":
+        """
+        A static method that converts a list of torch Tensors into an AgentAction using the ActionSpec.
+        This will change (and may be removed) in the ActionModel.
+        """
+        continuous: torch.Tensor = None
+        discrete: List[torch.Tensor] = None  # type: ignore
+        _offset = 0
+        if action_spec.continuous_size > 0:
+            continuous = tensor_list[0]
+            _offset = 1
+        if action_spec.discrete_size > 0:
+            discrete = tensor_list[_offset:]
+        return AgentAction(continuous, discrete)
+
+    @staticmethod
+    def from_dict(buff: Dict[str, np.ndarray]) -> "AgentAction":
+        """
+        A static method that accesses continuous and discrete action fields in an AgentBuffer
+        and constructs the corresponding AgentAction from the retrieved np arrays.
+        """
+        continuous: torch.Tensor = None
+        discrete: List[torch.Tensor] = None  # type: ignore
+        if "continuous_action" in buff:
+            continuous = ModelUtils.list_to_tensor(buff["continuous_action"])
+        if "discrete_action" in buff:
+            discrete_tensor = ModelUtils.list_to_tensor(
+                buff["discrete_action"], dtype=torch.long
+            )
+            discrete = [
+                discrete_tensor[..., i] for i in range(discrete_tensor.shape[-1])
+            ]
+        return AgentAction(continuous, discrete)
+
+
+class ActionLogProbs(NamedTuple):
+    """
+    A NamedTuple containing the tensor for continuous log probs and list of tensors for
+    discrete log probs of individual actions as well as all the log probs for an entire branch.
+    Utility functions provide numpy <=> tensor conversions to be used by the optimizers.
+    :param continuous_tensor: Torch tensor corresponding to log probs of continuous actions
+    :param discrete_list: List of Torch tensors each corresponding to log probs of discrete actions
+    :param all_discrete_list: List of Torch tensors each corresponding to all log probs of
+    a discrete action branch
```

Comment:
Let's try to make this more explicit : 
```suggestion
    :param discrete_list: List of Torch tensors each corresponding to log probs of the discrete actions that were sampled.
    :param all_discrete_list: List of Torch tensors each corresponding to all log probs of
    a discrete action branch, even the discrete actions that were not sampled. all_discrete_list is a list of Tensors,
    each Tensor corresponds to one discrete branch log probabilities. 
```